core/src/main/java/com/linecorp/armeria/internal/metric/CaffeineMetricSupport.java

```diff
@@ -47,63 +59,145 @@ public static void setup(MeterRegistry registry, MeterId id, Cache<?, ?> cache)
+                                  func(EVICTION_COUNT, ref -> ref.cacheStats.evictionCount()));
+            parent.more().counter(id.name("evictionWeight"), id.tags(), this,
+                                  func(EVICTION_WEIGHT, ref -> ref.cacheStats.evictionWeight()));
+            parent.gauge(id.name("estimatedSize"), id.tags(), this,
+                         func(null, ref -> ref.estimatedSize));
+        }
+
+        void add(Cache<?, ?> cache, Ticker ticker) {
+            synchronized (cacheRefs) {
+                for (CacheReference ref : cacheRefs) {
+                    if (ref.get() == cache) {
+                        // Do not aggregate more than once for the same instance.
+                        return;
+                    }
+                }
+
+                cacheRefs.add(new CacheReference(cache, ticker));
+            }
 
-            if (cache instanceof LoadingCache) {
+            if (cache instanceof LoadingCache && hasLoadingCache.compareAndSet(false, true)) {
+                // Add the following meters only for LoadingCache and only once.
                 final String loads = id.name("loads");
+
                 parent.more().counter(loads, id.tags("result", "success"), this,
-                                      new CacheStatFunction(() -> cacheStats.loadSuccessCount()));
+                                      func(LOAD_SUCCESS_COUNT, ref -> ref.cacheStats.loadSuccessCount()));
                 parent.more().counter(loads, id.tags("result", "failure"), this,
-                                      new CacheStatFunction(() -> cacheStats.loadFailureCount()));
-
-                final DoubleSupplier totalLoadTimeSupplier;
-                totalLoadTimeSupplier = () -> cacheStats.totalLoadTime();
+                                      func(LOAD_FAILURE_COUNT, ref -> ref.cacheStats.loadFailureCount()));
                 parent.more().counter(id.name("loadDuration"), id.tags(), this,
-                                      new CacheStatFunction(totalLoadTimeSupplier));
+                                      func(TOTAL_LOAD_TIME, ref -> ref.cacheStats.totalLoadTime()));
             }
+        }
 
-            parent.more().counter(id.name("evictions"), id.tags(), this,
-                                  new CacheStatFunction(() -> cacheStats.evictionCount()));
-            parent.more().counter(id.name("evictionWeight"), id.tags(), this,
-                                  new CacheStatFunction(() -> cacheStats.evictionWeight()));
-            parent.gauge(id.name("estimatedSize"), id.tags(), this,
-                         new CacheStatFunction(() -> estimatedSize));
+        private ToDoubleFunction<CaffeineMetrics> func(@Nullable Type type,
+                                                       ToDoubleFunction<CacheReference> valueFunction) {
+            return value -> {
+                double sum = 0;
+                synchronized (cacheRefs) {
```

Comment:
Is this called every time the cache metrics are updated? Seems like a lot of overhead on the normal case to satisfy a minor use case.

Reply:
It's only called when metrics are collected. Which shouldn't be very often, like once a minute?